auth/jwt-token-manager.ts

```diff
@@ -0,0 +1,169 @@
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. a) If this class is managing tokens and does not yet have one, make a request for one
+   *    b) If this class is managing tokens and the token has expired, request a new one
+   * 3. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo[this.tokenName] || this.isTokenExpired()) {
+      // 2. request a new token
+      this.requestToken((err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo[this.tokenName]);
+      });
+    } else {
+      // 3. use valid, sdk-managed token
+      return cb(null, this.tokenInfo[this.tokenName]);
+    }
+  }
+
+  /**
+   * Set a self-managed access token.
+   * The access token should be valid and not yet expired.
+   *
+   * By using this method, you accept responsibility for managing the
+   * access token yourself. You must set a new access token before this
+   * one expires. Failing to do so will result in authentication errors
+   * after this token expires.
+   *
+   * @param {string} accessToken - A valid, non-expired access token
+   * @returns {void}
+   */
+  public setAccessToken(accessToken: string): void {
+    this.userAccessToken = accessToken;
+  }
+
+  /**
+   * Request a JWT using an API key.
+   *
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  protected requestToken(cb: Function): void {
+    cb(null, 'token');
+  }
+
+  /**
+   * Compute and return a Basic Authorization header from a username and password.
+   *
+   * @param {string} username - The username or client id
+   * @param {string} password - The password or client secret
+   * @returns {string}
+   */
+  protected computeBasicAuthHeader(username, password): string {
+    const encodedCreds = Buffer.from(`${username}:${password}`).toString('base64');
+    return `Basic ${encodedCreds}`;
+  }
+
+  /**
+   * Check if currently stored token is expired.
+   *
+   * Using a buffer to prevent the edge case of the
+   * token expiring before the request could be made.
+   *
+   * The buffer will be a fraction of the total TTL. Using 80%.
+   *
+   * @private
+   * @returns {boolean}
+   */
+  private isTokenExpired(): boolean {
+    const { timeToLive, expireTime } = this;
+
+    if (!timeToLive || !expireTime) {
+      return true;
+    }
+
+    const fractionOfTtl = 0.8;
+    const currentTime = getCurrentTime();
+    const refreshTime = expireTime - (timeToLive * (1.0 - fractionOfTtl));
+    return refreshTime < currentTime;
```

Comment:
Probably shouldn't call this "refreshTime" here, since we've dropped the refresh logic.

Reply:
I hear you but am not sure what to call it. Going with `timeForNewToken` for now. Let me know if you have other ideas!